docs/articles/immutability-in-js.md

```diff
@@ -0,0 +1,108 @@
+
+## EcmaScript
+
+EcmaScript provides utilities to keep our data immutable. Arrays and objects APIs contain methods to create copies and even prevent instances from being updated. More recently EcmaScript introduced a new syntaxes to create copies of objects and arrays.
+
+### Object.assign
+
+We want to add a name property in our object.
+
+https://gist.github.com/frinyvonnick/a281f15288d3ab90ce5354f01df52ccf
+
+We can do it with `Object.assign` and a little trick. Basically it copies all properties from an object into another one, thus it mutates the target object. We use a small trick, passing an empty object as first parameter, which creates a new reference.
+
+https://gist.github.com/frinyvonnick/98878e45d6a8ccb4618238dbec175ae9
+
+We now have a new object with our new name property and a commonNames property remain unchanged. With this method you can create/overwrite multiple properties at the same time.
+
+### Array.concat
+
+Now let's do it with an array. We want to add two new elements in an array in an immutable way.
+
+https://gist.github.com/frinyvonnick/461d4dbad6bff965a60b734f86619faa
+
+Unlike `Array.push`, `Array.concat` does not mutate our array. Instead, it returns a new array.
+
+https://gist.github.com/frinyvonnick/3959e7972688a41055671bceb6228056
+
+This method is flexible. It takes as many elements as you want. They can be either values or arrays.
+
+### Object.freeze
+
+`Object.freeze` isn't really familiar. It lets you make an object immutable ! It prevents every type of mutation (creation, modification, deletion) induced by the use of setters.
+
+https://gist.github.com/frinyvonnick/d1dd1cbf2924b2428b07713d1ad2adf7
+
+We will try to delete name property after object has been frozen.
+
+https://gist.github.com/frinyvonnick/acc86d1c2fe14c23cd6363946cc253da
+
+Reallocation isn't necessary since the object passed as parameter has been made immutable by `Object.freeze`. This method has two available modes:
+
+- A non-strict mode that does not apply mutations
+- A strict mode that throws a `TypeError` if you try to apply mutations
+
+Be careful, it is not recursive. Our property commonNames isn't immutable.
+
+### Spread operator
+
+The spread operator syntax has been introduced in ES2015 for arrays and in ES2018 for objects. It copies all properties of a given object into a new object literal.
+
+https://gist.github.com/frinyvonnick/61908a0e1bdcd97ba965570796c0c665
+
+With arrays, it copies all values of an array into a new array.
+
+https://gist.github.com/frinyvonnick/7ec4cd039fd79f16864546c51b18ab95
+
+Easily readable and unified between arrays and objects, it replaces nicely assign and concat. It is possible to spread multiple arrays and objects in a same literal.
```

Comment:
J'aime pas trop la formule
Easily readable and unified between arrays and objects, it replaces nicely assign and concat => it replaces nicely `assign` and `concat` and from a developer perspective, it is easily readable and creates a convention between arrays and objects.